pkgs/build-support/node/build-npm-package/default.nix

```diff
@@ -0,0 +1,37 @@
+{ lib, stdenv, fetchNpmDeps, callPackage, nodejs }:
+
+{ name ? "${args.pname}-${args.version}"
+, src ? null
+, srcs ? null
+, sourceRoot ? null
+, patches ? [ ]
+, npmDepsHash ? ""
+, nativeBuildInputs ? [ ]
+, buildInputs ? [ ]
+, makeCacheWritable ? false
+, ...
+} @ args:
+
+let
+  npmDeps = fetchNpmDeps {
+    inherit src srcs sourceRoot patches;
+    name = "${name}-npm-deps";
+    hash = npmDepsHash;
+  };
+
+  # TODO: how do we expose these like Rust? `npmHooks`?
+  inherit (callPackage ./hooks { inherit nodejs; }) npmConfigHook npmBuildHook npmInstallHook;
+in
+stdenv.mkDerivation (args // {
+  inherit npmDeps;
+
+  nativeBuildInputs = nativeBuildInputs ++ [ nodejs npmConfigHook npmBuildHook npmInstallHook ];
+  buildInputs = buildInputs ++ [ nodejs ];
+
+  strictDeps = true;
+
+  nodeVersion = nodejs.version;
+  nodeVersionMajor = lib.versions.major nodejs.version;
```

Comment:
I wonder if these should be substitutions instead of environment variables, or maybe the Node source should be moved. (These were originally substitutions, but I moved them for... I don't know why.)

Reply:
After sitting on this for a bit, my current thoughts on this are that these are implementation details of the hook that we don't want to be configurable, so they should be substitutions.